utilities/qneg.F90

```diff
@@ -0,0 +1,509 @@
+    allocate(qneg3_warn_num(num_constituents, num3_bins), stat=ierr)
+    call check_allocate(ierr, subname, 'qneg3_warn_num')
+    allocate(qneg3_warn_worst(num_constituents, num3_bins), stat=ierr)
+    call check_allocate(ierr, subname, 'qneg3_warn_worst')
+    qneg3_warn_num = 0
+    qneg3_warn_worst = worst_reset
+
+    select case(trim(print_qneg_warn))
+       case('summary')
+          collect_stats = .true.
+          timestep_reset = .false.
+       case('timestep')
+          collect_stats = .true.
+          timestep_reset = .true.
+       case('off')
+          collect_stats = .false.
+          timestep_reset =.false.
+       case default
+          !do nothing
+!          call endrun(sub//"FATAL: '"//trim(print_qneg_warn)//"' is not a    &
+!             valid value for print_qneg_warn")
+    end select
+
+    !Set qneg_initialized to .true.
+    qneg_initialized = .true.
+
+  end subroutine qneg_init
+
+!  subroutine calc_cnst_out()
+!    use cam_history, only: hist_fld_active, history_initialized
+!    integer :: index
+!
+!    if (history_initialized()) then
+!       ! to protect against routines that call qneg3 too early
+!       do index = 1, num_diag_fields
+!          cnst_outfld(index) = hist_fld_active(trim(diag_names(index)))
+!       end do
+!       cnst_out_calc = .true.
+!    end if
+!
+!  end subroutine calc_cnst_out
+
+  integer function find_index3(nam) result(index)
+    ! Find a valid or new index for 'nam' entries
+    character(len=*),  intent(in) :: nam
+
+    integer                      :: i
+
+    index = -1
+    do i = 1, num3_bins
+       if (trim(nam) == trim(qneg3_warn_labels(i))) then
+          ! We found this entry, return its index
+          index = i
+          exit
+       else if (len_trim(qneg3_warn_labels(i)) == 0) then
+          ! We have run out of known entries, use a new one and reset its stats
+          qneg3_warn_labels(i) = nam
+          index = i
+          call reset_stats(qneg3_warn_num(:, index), qneg3_warn_worst(:,index))
+          exit
+       end if
+    end do
+  end function find_index3
+
+!  integer function find_index4(nam) result(index)
+!    ! Find a valid or new index for 'nam' entries
+!    character(len=*),  intent(in) :: nam
+!
+!    integer                      :: i
+!
+!    index = -1
+!    do i = 1, num4_bins
+!       if (trim(nam) == trim(qneg4_warn_labels(i))) then
+!          ! We found this entry, return its index
+!          index = i
+!          exit
+!       else if (len_trim(qneg4_warn_labels(i)) == 0) then
+!          ! We have run out of known entries, use a new one and reset its stats
+!          qneg4_warn_labels(i) = nam
+!          index = i
+!          call reset_stats(qneg4_warn_num(index), qneg4_warn_worst(index))
+!          exit
+!       end if
+!    end do
+!  end function find_index4
+
+!> \section arg_table_qneg_run Argument Table
+!! \htmlinclude qneg_run.html
+  subroutine qneg_run (subnam, ncol, ncold, lver, num_constituents, qmin, q)
```

Comment:
Should have `errcode` and `errmsg` (initialize them to zero and empty string respectively).